File: examples/nodejs-agent/agent.js
#!/usr/bin/env node
/**
 * Node.js MailGoat example agent.
 * Polls inbox, parses messages, and sends automated responses.
 */

import { execFile } from 'node:child_process';
import { promises as fs } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { promisify } from 'node:util';

const execFileAsync = promisify(execFile);

const POLL_SECONDS = Number(process.env.POLL_SECONDS || 300);
const STATE_FILE = (process.env.STATE_FILE || path.join(os.homedir(), '.mailgoat-node-agent-state.json')).replace('~', os.homedir());
const RESPONSE_TAG = process.env.RESPONSE_TAG || 'auto-node-agent';

async function runMailgoat(args) {
  const { stdout, stderr } = await execFileAsync('mailgoat', [...args, '--json']);
  if (stderr && stderr.trim()) {
    // mailgoat may print warnings; don't fail on non-fatal stderr.
  }
  return stdout?.trim() ? JSON.parse(stdout) : null;
}

async function loadState() {
  try {
    const raw = await fs.readFile(STATE_FILE, 'utf8');
    const payload = JSON.parse(raw);
    return new Set(payload.processedIds || []);
  } catch {
    return new Set();
  }
}

async function saveState(ids) {
  const payload = { processedIds: [...ids].sort() };
  await fs.writeFile(STATE_FILE, `${JSON.stringify(payload, null, 2)}\n`, 'utf8');
}

function summarizeMessage(message) {
  const body = String(message.body || '').trim().replace(/\s+/g, ' ');
  return body.length > 180 ? `${body.slice(0, 180)}...` : body;
}

function buildReply(message) {
  const from = message.from || 'unknown@local';
  const subject = message.subject || '(no subject)';
  const summary = summarizeMessage(message) || '(empty body)';

  return {
    to: from,
    subject: `Re: ${subject}`,
    body: [
      `Hello ${from},`,
      '',
      'Your email has been processed by the Node.js example agent.',
      `Summary: ${summary}`,
      '',
      'This is an automated response.',
    ].join('\n'),
  };
}

async function processInbox() {
  const state = await loadState();
  const messages = await runMailgoat(['inbox', 'list', '--limit', '50']);

  if (!Array.isArray(messages)) {
    console.error('Unexpected inbox payload (expected array).');
    return;
  }

  for (const row of messages) {
    const id = String(row.id || '').trim();
    if (!id || state.has(id)) continue;

    const message = await runMailgoat(['read', id]);
    if (!message?.from) {
      state.add(id);
      continue;
    }

    const reply = buildReply(message);
    await runMailgoat([
      'send',
      '--to', reply.to,
      '--subject', reply.subject,
      '--body', reply.body,
      '--tag', RESPONSE_TAG,
    ]);

    state.add(id);
    console.log(`Replied to ${reply.to} for ${id}`);
  }

  await saveState(state);
}

async function main() {
  console.log(`Node.js agent started. Poll interval: ${POLL_SECONDS}s`);
  for (;;) {
    try {
      await processInbox();
    } catch (error) {
      console.error('Agent loop error:', error.message);
    }
    await new Promise((resolve) => setTimeout(resolve, POLL_SECONDS * 1000));
  }
}

main();

File: examples/simple-agent/agent.py
#!/usr/bin/env python3
"""
Simple MailGoat agent.

Polls local MailGoat inbox cache every N seconds, reads new messages,
and sends automated responses via `mailgoat send`.
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
import time
from pathlib import Path
from typing import Any

POLL_SECONDS = int(os.getenv("POLL_SECONDS", "300"))
STATE_PATH = Path(os.getenv("STATE_FILE", str(Path.home() / ".mailgoat-simple-agent-state.json")))
REPLY_PREFIX = os.getenv("REPLY_PREFIX", "[Auto-Reply]")
FROM_HINT = os.getenv("MAILGOAT_FROM", "")


def run_mailgoat(args: list[str]) -> Any:
    """Run mailgoat command and parse JSON output."""
    cmd = ["mailgoat", *args, "--json"]
    proc = subprocess.run(cmd, capture_output=True, text=True)
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or proc.stdout.strip() or "mailgoat command failed")
    out = proc.stdout.strip()
    return json.loads(out) if out else None


def load_state() -> set[str]:
    if not STATE_PATH.exists():
        return set()
    try:
        payload = json.loads(STATE_PATH.read_text(encoding="utf-8"))
        return set(payload.get("processed_ids", []))
    except Exception:
        return set()


def save_state(processed_ids: set[str]) -> None:
    STATE_PATH.parent.mkdir(parents=True, exist_ok=True)
    STATE_PATH.write_text(
        json.dumps({"processed_ids": sorted(processed_ids)}, indent=2),
        encoding="utf-8",
    )


def build_reply_subject(original_subject: str) -> str:
    subject = original_subject.strip() if original_subject else "(no subject)"
    if subject.lower().startswith("re:"):
        return f"{REPLY_PREFIX} {subject}"
    return f"{REPLY_PREFIX} Re: {subject}"


def build_reply_body(sender: str, subject: str) -> str:
    lines = [
        f"Hello {sender or 'there'},",
        "",
        "Your email was received by the simple MailGoat agent.",
        "This is an automated acknowledgment while a human reviews your message.",
        "",
        f"Original subject: {subject or '(no subject)'}",
        "",
        "- MailGoat Simple Agent",
    ]
    return "\n".join(lines)


def process_messages() -> None:
    processed = load_state()

    inbox = run_mailgoat(["inbox", "list", "--limit", "50"])
    if not isinstance(inbox, list):
        print("Unexpected inbox payload, expected JSON array", file=sys.stderr)
        return

    for entry in inbox:
        msg_id = str(entry.get("id", "")).strip()
        if not msg_id or msg_id in processed:
            continue

        details = run_mailgoat(["read", msg_id])
        sender = str(details.get("from", "")).strip()
        subject = str(details.get("subject", "")).strip()

        if not sender:
            processed.add(msg_id)
            continue

        send_args = [
            "send",
            "--to",
            sender,
            "--subject",
            build_reply_subject(subject),
            "--body",
            build_reply_body(sender, subject),
        ]
        if FROM_HINT:
            send_args.extend(["--from", FROM_HINT])

        run_mailgoat(send_args)
        processed.add(msg_id)
        print(f"Replied to {sender} for message {msg_id}")

    save_state(processed)


def main() -> None:
    print(f"Simple agent started. Poll interval: {POLL_SECONDS}s")
    while True:
        try:
            process_messages()
        except Exception as exc:
            print(f"Agent iteration failed: {exc}", file=sys.stderr)
        time.sleep(POLL_SECONDS)


if __name__ == "__main__":
    main()
